chia/wallet/coin_selection.py

```diff
@@ -0,0 +1,138 @@
+    """
+    if exclude is None:
+        exclude = []
+
+    if amount > spendable_amount:
+        error_msg = (
+            f"Can't select amount higher than our spendable balance.  Amount: {amount}, spendable: {spendable_amount}"
+        )
+        log.warning(error_msg)
+        raise ValueError(error_msg)
+
+    log.info(f"About to select coins for amount {amount}")
+
+    sum_spendable_coins = 0
+    valid_spendable_coins: List[Coin] = []
+
+    for coin_record in spendable_coins:  # remove all the useless coins.
+        if coin_record.coin.name() in unconfirmed_removals:
+            continue
+        if coin_record.coin in exclude:
+            continue
+        valid_spendable_coins.append(coin_record.coin)
+        sum_spendable_coins += coin_record.coin.amount
+
+    # This happens when we couldn't use one of the coins because it's already used
+    # but unconfirmed, and we are waiting for the change. (unconfirmed_additions)
+    if sum_spendable_coins < amount:
+        raise ValueError(
+            "Can't make this transaction at the moment. Waiting for the change from the previous transaction."
+        )
+
+    # Sort the coins by amount
+    valid_spendable_coins.sort(reverse=True, key=lambda r: r.amount)
+
+    # check for exact 1 to 1 coin match.
+    exact_match_coin: Optional[Coin] = check_for_exact_match(valid_spendable_coins, uint64(amount))
+    if exact_match_coin:
+        log.info(f"selected coin with an exact match: {exact_match_coin}")
+        return {exact_match_coin}
+
+    # Check for an exact match with all of the coins smaller than the amount.
+    # If we have more, smaller coins than the amount we run the next algorithm.
+    smaller_coin_sum = 0  # coins smaller than target.
+    smaller_coins: Set[Coin] = set()
+    for coin in valid_spendable_coins:
+        if coin.amount < amount:
+            smaller_coin_sum += coin.amount
+            smaller_coins.add(coin)
+    if smaller_coin_sum == amount:
+        log.info(f"Selected all smaller coins because they equate to an exact match of the target.: {smaller_coins}")
+        return smaller_coins
+    elif smaller_coin_sum < amount:
+        smallest_coin: Optional[Coin] = find_smallest_coin(valid_spendable_coins, amount, max_coin_amount)
+        assert smallest_coin is not None
+        log.info(f"Selected closest greater coin: {smallest_coin.name()}")
+        return {smallest_coin}
+    else:
+        best_coin_set = knapsack_coin_algorithm(smaller_coins, amount, max_coin_amount)
+        assert best_coin_set is not None
+        log.info(f"Selected coins from knapsack algorithm: {best_coin_set}")
+        return best_coin_set
+
+
+# These algorithms were based off of the algorithms in:
+# https://murch.one/wp-content/uploads/2016/11/erhardt2016coinselection.pdf
+
+# we use this to check if one of the coins exactly matches the target.
+def check_for_exact_match(coin_list: List[Coin], target: uint64) -> Optional[Coin]:
+    for coin in coin_list:

```

Comment:
The list is sorted, so we could do better than iterating the list if we wanted more performance (binary search anyone - unfortunately since sorted descending, I don't think we can use bisect_left directly). This list of spendable coins could be quite large.

Reply:
I am just going to remove this, as it should already be properly sorted.